client/virtual_channel_settlement.go

```diff
@@ -0,0 +1,188 @@
+			FinalStateSignatures: virtual.machine.CurrentTX().Sigs,
+		}
+	})
+
+	return errors.WithMessage(err, "update parent channel")
+}
+
+func (c *Client) handleVirtualChannelSettlementProposal(
+	parent *Channel,
+	prop *virtualChannelSettlementProposal,
+	responder *UpdateResponder,
+) {
+	err := c.validateVirtualChannelSettlementProposal(parent, prop)
+	if err != nil {
+		c.rejectProposal(responder, err.Error())
+	}
+
+	ctx, cancel := context.WithTimeout(c.Ctx(), virtualChannelFundingTimeout)
+	defer cancel()
+
+	err = c.settlementWatcher.Await(ctx, prop)
+	if err != nil {
+		c.rejectProposal(responder, err.Error())
+	}
+
+	c.acceptProposal(responder)
+}
+
+func (c *Client) validateVirtualChannelSettlementProposal(
+	parent *Channel,
+	prop *virtualChannelSettlementProposal,
+) error {
+	// Validate parameters.
+	if prop.ChannelParams.ID() != prop.FinalState.ID {
+		return errors.New("invalid parameters")
+	}
+
+	// Validate signatures.
+	for i, sig := range prop.FinalStateSignatures {
+		ok, err := channel.Verify(
+			prop.ChannelParams.Parts[i],
+			&prop.ChannelParams,
+			&prop.FinalState,
+			sig,
+		)
+		if err != nil {
+			return err
+		} else if !ok {
+			return errors.New("invalid signature")
+		}
+	}
+
+	// Validate allocation.
+
+	// Assert equal assets.
+	if err := channel.AssetsAssertEqual(parent.state().Assets, prop.FinalState.Assets); err != nil {
+		return errors.WithMessage(err, "assets do not match")
+	}
+
+	// Assert contained before and matching funds
+	subAlloc, containedBefore := parent.state().SubAlloc(prop.ChannelParams.ID())
+	if !containedBefore || !subAlloc.BalancesEqual(prop.FinalState.Sum()) {
+		return errors.New("virtual channel not allocated")
+	}
+
+	// Assert not contained after
+	_, containedAfter := prop.State.SubAlloc(prop.ChannelParams.ID())
+	if containedAfter {
+		return errors.New("virtual channel must not be de-allocated after update")
+	}
+
+	// Assert correct balances
+	virtual := transformBalances(prop.FinalState.Balances, parent.state().NumParts(), subAlloc.IndexMap)
+	correctBalances := parent.state().Balances.Add(virtual).Equal(prop.State.Balances)
+	if !correctBalances {
+		return errors.New("invalid balances")
+	}
+
+	return nil
+}
+
+func (c *Client) matchSettlementProposal(a, b interface{}) (ok bool) {
+	// Cast.
+	inputs := []interface{}{a, b}
+	props := make([]*virtualChannelSettlementProposal, len(inputs))
+	for i, x := range inputs {
+		var prop *virtualChannelSettlementProposal
+		prop, ok = x.(*virtualChannelSettlementProposal)
+		if !ok {
+			return
+		}
+		props[i] = prop
+	}
+
+	prop0 := props[0]
+
+	for _, prop := range props {
+		// Check final state.
+		ok = prop.FinalState.Equal(&prop0.FinalState) == nil
+		if !ok {
+			return
+		}
+	}
+
+	// Store settlement state and signature.
+	virtual, err := c.Channel(prop0.FinalState.ID)
+	if err != nil {
+		return false
+	}
+	if err = virtual.machine.Update(c.Ctx(), &prop0.FinalState, hubIndex); err != nil {
```

Comment:
Does the state machine allow the skipping of version numbers? Did you test to do a couple of direct updates in the virtual channels so that the version is at least 2 when settling?

Reply:
Nice catch. I only did exactly one update in the test 🙈 Now added an additional transaction to the test and a ForceUpdate method.